bitmark-references/bitmark-node-software/mining.md

```diff
@@ -0,0 +1,117 @@
+            |              |  ---------------->  |               |
+            |   bitmarkd   |                     |   recorderd   |
+            |              |  <----------------  |               |
+            +--------------+  zero mq protocol   +---------------+
+
+The hashing procedure relates to `bitmarkd` and `recorderd` only. `recorderd` is program to perform hashing, it receives jobs from `bitmarkd` and tries to find hashes possibly meets hashing criteria, if any possible hash is found, `recorderd` sends message back to `bitmarkd` to ask for validation.
+
+When `bitmarkd` receives special string from `recorderd`, `bitmarkd` validates hashes from this special string and returns validation result to `recorderd`.
+
+If `bitmarkd` receives no valid hashes from other nodes or from `recorderd`, `bitmarkd` sends hashing task to `recorderd` periodically (currently every 1 minute), until valid hash is found or received.
+
+## Hashing
+
+Hashing is a behavior to convert any form of data into a unique string of text, if a hashing algorithm is well designed, it should be hard to guess original data from generated output.
+
+Hashing algorithm used in bitmark is `argon2`, which is a memory hard algorithm that can resist from GPU or ASIC hardware computation. Length of a hash is 256 bits long.
+
+Many parameters relates to hashing, if `argon2` executable is installed, following command line can demonstrate the hasing process:
+
+```
+printf '%s' 'hello world' | argon2 'hello world' -d -l 32 -m 17 -t 4 -p 1 -r | awk '{for(i=length($1);i>0;i-=2)x=x substr($1,i-1,2);print x}'
+
+f8a17bc25cb53e848e2d09811ade4b8a037f628443661b88611faf5d9a5a1f33
+```
+
+A block hash contains following data:
+
+1. block record version
+
+    blockchain header version
+
+1. transacion count
+
+    number of transactions in the block
+
+1. merkle root
+
+    the hash of all transaction hashes inside the block, if verified transactions are less than block allowable limit (9999), then all verified transactions will be put into a block. If verified transactions above block allowable limit, random transactions will be selected
+
+1. timestamp
+
+    block generated time
+
+1. difficulty
+
+    hashing difficulty
+
+1. nonce
+
+    random string to make hash fits difficuty
+
+## Difficulty
+
+Difficulty is a number to decide which hash is valid, higher difficulty means harder to find a valid hash. A hash meets difficulty means hash vaue is less than or equal to difficulty value.
+
+The hash is considered as a 256 bit fixed point value composeda as following two parts:
+
+    [8 bits] [248 bits]
+    exponent mantissa
+
+Difficulty value is encoded by 64 bits, including 8 bits of exponent and 56 bits of mantissa.
+
+    [56 bits] [8 bits]
+    mantissa  exponent = 64 bits unsigned value
+
+Mantissa is normalized so that most significat bits is one and can be droppped leaving 56 bits to store, itt is right shifted by exponent +8 from the most significant bit.
```

Comment:
From:
Mantissa is normalized so that most significat bits is one and can be droppped leaving 56 bits to store, itt is right shifted by exponent +8 from the most significant bit.

To:
The mantissa bits are normalized, the most significant bit is 1 and can be dropped, leaving the 56 bits to store; these bits are right shifted by exponent +8 from the most significant bit.

Reply:
done